molgenis-data/src/main/java/org/molgenis/data/listeners/EntityListenersService.java

```diff
@@ -0,0 +1,105 @@
+package org.molgenis.data.listeners;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.SetMultimap;
+import org.molgenis.data.Entity;
+import org.molgenis.data.MolgenisDataException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.PreDestroy;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+
+@Service
+public final class EntityListenersService
+{
+	private final Logger LOG = LoggerFactory.getLogger(EntityListenersService.class);
+	private final ConcurrentHashMap<String, SetMultimap<Object, EntityListener>> entityListeners = new ConcurrentHashMap<>();
+
+	/**
+	 * Register a repository to the entity listeners service once
+	 *
+	 * @param repoFullName
+	 * @return
+	 */
+	void register(String repoFullName)
+	{
+		if (!entityListeners.containsKey(requireNonNull(repoFullName)))
+		{
+			entityListeners.put(repoFullName, HashMultimap.create());
+		}
+	}
+
+	Stream<Entity> updateEntities(String repoFullName, Stream<Entity> entities)
+	{
+		verifyRepoRegistered(repoFullName);
+		SetMultimap<Object, EntityListener> entityListeners = this.entityListeners.get(repoFullName);
+		synchronized (entityListeners)
```

Comment:
I think it should be possible to concurrently read from the listeners, perhaps you need to create an immutable copy or so. But that'll help prevent synchronization issues. This synchronized block effectively locks the repository.